### PR TITLE
Represent string constants as strings

### PR DESCRIPTION
**Pull request by widlarizer**
Currently, a `RTLIL::Const` representing a string will use `std::vector<RTLIL::State>`, encoding the string into a vector of individual bits. This increases the memory consumption of string attributes, such as `src` and `hdlname`, by a factor of 8.

As of this PR, the structure now holds data represented with ~~an `std::variant<bitvectype, std::string>` named `backing`~~ a union. The `bits` vector is no longer accessible and instead `std::vector<RTLIL::State>& bits()` has to be used. When a Const is constructed from a string, it is represented with the string alternative. When its `bits()` method is used, the string is encoded into the bit vector alternative as prior to this PR, see `Const::bitvectorize()`. This is similar to how `RTLIL::SigSpec` "unpacks" into bits.

In the initial implementation, when `bits()` were being accessed, I asserted that the Const isn't represented with a string. This found some places in Yosys that access string attributes bit-by-bit. As a result, this PR comes with bonus content:

+ `kernel/cellaigs.cc`
+ `frontends/ast/ast.cc`
+ `Const::pretty_fmt`
+ cell checker `is_param` that doesn't try return bits

It's possible to split these now necessary changes out to a separate PR, but it will be heavy on git conflicts. Please advise.

This PR does not affect constant `SigChunk`s which also have an `std::vector<RTLIL::State>` each.

- [X] clean my desk (TODO, log_debug, did_something_hook, unwrap `std::get` from `assert_get_bits` etc)
- [x] ~~add some way of debugging how much unpacking we're doing~~ requires unrelated changes in followup PR

Please comment what memory usage changes you see with this PR. I've observed memory savings of 0-20% depending on the design with negligible runtime improvements. The open source Verilog frontend often dominates memory usage, so flows that use Verific may more readily show these improvements.

### CODE REVIEW

File: kernel/rtlil.cc [hunk 1]
```diff
@@ -203,93 +203,171 @@ const pool<IdString> &RTLIL::builtin_ff_cell_types() {
...
-	if (is_signed && bits.back() == State::S1)
-		for (size_t i = bits.size(); i < 32; i++)
+	if (is_signed && bv.back() == State::S1)
+		for (size_t i = bv.size(); i < 32; i++)
 			ret |= 1 << i;
 	return ret;
 }
 
-std::string RTLIL::Const::as_string() const
+std::string RTLIL::Const::as_string(std::string any) const
```
Comment (povik) [inline kernel/rtlil.cc:358]: What's up with this `any` parameter? This looks unrelated to the backing change

File: kernel/rtlil.cc [hunk 2]
```diff
@@ -344,9 +429,11 @@ std::string RTLIL::Const::decode_string() const
 
 bool RTLIL::Const::is_fully_zero() const
 {
+	bitvectorize();
+	bitvectype& bv = assert_get_bits("Const::decode_string");
```
Comment (povik) [inline kernel/rtlil.cc:433]: Why not just call `bits()` instead of repeating `bitvectorize(); bitvectype& bv = assert_get_bits(...)` everywhere?

File: kernel/rtlil.h [hunk 3]
```diff
@@ -673,19 +687,68 @@ struct RTLIL::Const
...
+		bitvectorize();
+		return assert_get_bits("Const::operator[]").at(index);
+	}
+	inline const RTLIL::State &operator[](int index) const {
+		bitvectorize();
+		return assert_get_bits("const Const::operator[]").at(index);
+	}
+	inline bitvectype::iterator begin() {
+		bitvectorize();
+		return assert_get_bits("Const bit iterator begin()").begin();
```
Comment (povik) [inline kernel/rtlil.h:746]: I think this is not constant, you can make changes through this iterator

File: kernel/rtlil.h [hunk 4]
```diff
@@ -695,25 +758,37 @@ struct RTLIL::Const
...
 	}
 
 	inline unsigned int hash() const {
 		unsigned int h = mkhash_init;
-		for (auto b : bits)
-			h = mkhash(h, b);
+
+		if(auto str = std::get_if<std::string>(&backing)) {
+			for (auto c : *str)
+				h = mkhash(h, c);
```
Comment (povik) [inline kernel/rtlil.h:786]: If you have two Consts, one uses the string backing but the other uses the bit vector backing, this doesn't seem to guarantee for their hashes to be the same when their bit content is the same

File: kernel/rtlil.cc [hunk 5]
```diff
@@ -203,93 +203,171 @@ const pool<IdString> &RTLIL::builtin_ff_cell_types() {
...
+		unsigned char ch = str[i];
+		for (int j = 0; j < 8; j++) {
+			b.push_back((ch & 1) != 0 ? State::S1 : State::S0);
+			ch = ch >> 1;
+		}
+	}
+	return b;
+}
+
+std::string RTLIL::Const::pretty_fmt() const {
```
Comment (povik) [inline kernel/rtlil.cc:320]: What are these new functions for?

File: kernel/rtlil.cc [hunk 6]
```diff
@@ -203,93 +203,171 @@ const pool<IdString> &RTLIL::builtin_ff_cell_types() {
...
+			return bv[i] < other_bv[i];
 	return false;
 }
 
 bool RTLIL::Const::operator ==(const RTLIL::Const &other) const
 {
-	return bits == other.bits;
+	const char* ctx = "operator==";
+	if (std::get_if<std::string>(&backing) != std::get_if<std::string>(&other.backing))
+		return decode_string() == other.decode_string();
```
Comment (povik) [inline kernel/rtlil.cc:276]: `decode_string()` can be equal for two Const even if their content isn't. Consider `Const(1, 1)` vs `Const(1, 8)`

File: kernel/rtlil.h (hunk 3 above)
Comment (widlarizer) [reply kernel/rtlil.h:746]: Yeah, I meant to refer to `RTLIL::Const`, not the C++ `const` keyword

File: kernel/rtlil.cc [hunk 7]
```diff
@@ -203,93 +203,171 @@ const pool<IdString> &RTLIL::builtin_ff_cell_types() {
...
+		if (bv[i] != other_bv[i])
+			return bv[i] < other_bv[i];
 	return false;
 }
 
 bool RTLIL::Const::operator ==(const RTLIL::Const &other) const
 {
-	return bits == other.bits;
+	const char* ctx = "operator==";
+	if (std::get_if<std::string>(&backing) != std::get_if<std::string>(&other.backing))
```
Comment (povik) [inline kernel/rtlil.cc:275]: This is a confusing conditional. The only way the two `get_if` can return the same pointer is if they both return null, or if the Const is being compared against itself, right?

File: kernel/rtlil.h (hunk 3 above)
Comment (povik) [reply kernel/rtlil.h:746]: OK, then we may drop Const from there as it's superfluous

File: kernel/rtlil.cc (hunk 2 above)
Comment (povik) [reply kernel/rtlil.cc:433]: Ah you have a TODO to get rid of this

File: backends/cxxrtl/cxxrtl_backend.cc [hunk 8]
```diff
@@ -864,7 +864,7 @@ struct CxxrtlWorker {
 		if (!module->has_attribute(ID(cxxrtl_template)))
 			return {};
 
-		if (module->attributes.at(ID(cxxrtl_template)).flags != RTLIL::CONST_FLAG_STRING)
```
Comment (povik) [inline backends/cxxrtl/cxxrtl_backend.cc:867]: Here we are giving the flag a semantical significance, I think we may want to remove this check

Comment (whitequark) [reply backends/cxxrtl/cxxrtl_backend.cc:867]: I'm going to reject this change.

Comment (whitequark) [reply backends/cxxrtl/cxxrtl_backend.cc:867]: As far as I'm concerned the `STRING` flag has semantic significance and CXXRTL treats it as such in many places (e.g. the types are going to be different and this will change the ABI). I don't really care if RTLIL claims there's no significance to it, fix RTLIL then.

Comment (povik) [reply backends/cxxrtl/cxxrtl_backend.cc:867]: With this PR we are removing the semantical significance of `STRING`, instead it serves to signify the backing used, which can change merely by using the wrong accessor on the Const object. We may not go through with it, but that's the plan. As an alternative we can add an extra flag which would signify the backing.

FWIW the `==` and `!=` operators on Const disregard the flags as of today, so we have an inconsistence either way.

File: kernel/rtlil.cc [hunk 9]
```diff
@@ -1034,6 +1135,14 @@ namespace {
 					cell->name.c_str(), cell->type.c_str(), __FILE__, linenr, buf.str().c_str());
 		}
 
+		void is_param(const RTLIL::IdString& name)
```
[the remark this answers is not in the document]
Comment (widlarizer) [reply kernel/rtlil.cc:1138]: It's related, having a version of this function that doesn't have a return value prevents unpacking every single string attribute to bits. Without this, the PR wouldn't have any positive impact because once you ran check, all Const would unpack

[the remark this answers is not in the document]
Comment (povik) [reply kernel/rtlil.cc:1138]: That's for a parameter not an attribute though

File: backends/cxxrtl/cxxrtl_backend.cc (hunk 8 above)
Comment (whitequark) [reply backends/cxxrtl/cxxrtl_backend.cc:867]: CXXRTL relies on the semantic significance so I'll not accept this PR until there is a way forward for it.

Comment (povik) [reply backends/cxxrtl/cxxrtl_backend.cc:867]: > With this PR we are removing the semantical significance of STRING, instead it serves to signify the backing used, which can change merely by using the wrong accessor on the Const object.

Looking at the code change that may not be true (but it's what I thought we were doing). So the backing seems decided by the state of the `std::variant<>` field, in which case I need to take back what I said about the `STRING` flag.

We still have a somewhat inconsistent state with the comparison operators ignoring the flags, but that may be exposed through to the user in some places and I am not sure we can change it.

Comment (whitequark) [reply backends/cxxrtl/cxxrtl_backend.cc:867]: Note also that this conditional isn't the only (nor the most important) place where this is done; the attribute set dumping code is the primary one. In particular the results of this transformation are used for ChipFlow tools and we rely on it being done correctly.

Comment (whitequark) [reply backends/cxxrtl/cxxrtl_backend.cc:867]: > We still have a somewhat inconsistent state with the comparison operators ignoring the flags, but that may be exposed through to the user in some places and I am not sure we can change it.

I guess this happens because in Verilog the string compares equal to the corresponding constant, and since Yosys is mainly used with Verilog nobody objected to this. I think that a good way to fix this issue woudl be to emit a warning (while comparing unequal) so that any such desired reliance can be found and retained with an explicit conversion.

File: kernel/rtlil.cc (hunk 1 above)
Comment (widlarizer) [reply kernel/rtlil.cc:358]: Yes, it's unrelated refactoring to adjacent code. I'm giving as_string a parameter that sets the representation for don't-care / any bits to deduplicate code that previously used the `bits()` interface on some strings. I think currently it only deduplicates it with `serialize_param_value` in `ast.cc` but there's other places where you could use this method now

Comment (povik) [reply kernel/rtlil.cc:358]: Ideally this would be a separate commit then. Also this would be easier to review if the `s/.bits/.bits()` change was a commit of its own, done upfront before we start making changes to the backing within `Const`.

File: kernel/rtlil.cc (hunk 5 above)
Comment (widlarizer) [reply kernel/rtlil.cc:320]: To move code in from elsewhere for deduplication and code reuse. This one is used in `Aig::Aig(Cell *cell)` now

File: kernel/rtlil.cc (hunk 7 above)
Comment (widlarizer) [reply kernel/rtlil.cc:275]: Good catch, forgot to cast to bool

File: kernel/rtlil.cc (hunk 5 above)
Comment (widlarizer) [reply kernel/rtlil.cc:320]: Right, previously, it was `name += stringf(":%d", p.second.as_int());` which would unpack string constants. We want to avoid that in all scenarios. Unpacking to bits is a safety measure, only because I can't guarantee it can't happen in some use case

Comment (povik) [reply kernel/rtlil.cc:320]: > which would unpack string constants. We want to avoid that in all scenarios.

Not here, these are the `WIDTH` parameters which are unlikely to have string backing in the first place

Comment (povik) [reply kernel/rtlil.cc:320]: We don't care about backing for parameters at all, at least not yet, they are almost never strings (the only exception I can think of is `FLAVOR` on `$check`; and parameters on user modules of course)

File: backends/cxxrtl/cxxrtl_backend.cc (hunk 8 above)
Comment (widlarizer) [reply backends/cxxrtl/cxxrtl_backend.cc:867]: Note: I wrote that not having seen the prior comments (github pls). I'll respond to it more properly soon when I have enough time to understand the situation some more

File: kernel/rtlil.cc (hunk 5 above)
Comment (widlarizer) [reply kernel/rtlil.cc:320]: Ah true, attributes are not covered by this. WIDTH parameters specifically still use `param` rather than `is_param` because their value is used. I still don't like enforcing non-bit-ness when we're only checking for param presence

Comment (povik) [reply kernel/rtlil.cc:308]: > WIDTH parameters specifically still use `param` rather than `is_param` because their value is used.

I was commenting on the cellaigs code which does `name += stringf(":%d", p.second.as_int());` which touches on the `WIDTH` params (`A_WIDTH` or `B_WIDTH`).

> I still don't like enforcing bitvec backing when we're only checking for param presence

Regarding the `is_param` vs `param` situation -- there aren't any gains to expect there, on the other hand any change we can shed off from this PR will make the review easier, which is important given the wide-reaching nature of it, for that reason I think it's better we drop that change.

File: backends/cxxrtl/cxxrtl_backend.cc (hunk 8 above)
Comment (povik) [reply backends/cxxrtl/cxxrtl_backend.cc:867]: > I'll respond to it more properly soon when I have enough time to understand the situation some more

The key takeaway is this: not to upset users we need the choice of backing for Const to be independent of whether the `STRING` flag is set, which I think is already the case with this PR.

**Comment by povik** [comment]
> Please comment what memory usage changes you see with this PR.

On the sky130hd/jpeg design from OpenROAD-flow-scripts, dumping the netlist just after the `techmap` command, I see the following difference in memory usage when loading in the RTLIL dump in a fresh Yosys process:
| | with `src` attributes | `src` stripped |
|---|---|---|
| baseline | 291 MB | 180 MB |
| patched | 221 MB | 172 MB |

File: kernel/rtlil.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (widlarizer) [reply kernel/rtlil.cc:1126]: Ok, removed

File: kernel/rtlil.h (hunk 3 above)
Comment (widlarizer) [reply kernel/rtlil.h]: This wrapper is removed and we lean on exceptions in the unexpected case of variant access problems instead

**Comment by povik** [comment]
I have tried the following patch to hashlib on top of the PR, and updated the table:

```
diff --git a/kernel/hashlib.h b/kernel/hashlib.h
index 6b880f3a6..0e839e73e 100644
--- a/kernel/hashlib.h
+++ b/kernel/hashlib.h
@@ -190,7 +190,7 @@ inline int hashtable_size(int min_size)
 {
        // Primes as generated by https://oeis.org/A175953
        static std::vector<int> zero_and_some_primes = {
-               0, 23, 29, 37, 47, 59, 79, 101, 127, 163, 211, 269, 337, 431, 541, 677,
+               0, 3, 23, 29, 37, 47, 59, 79, 101, 127, 163, 211, 269, 337, 431, 541, 677,
                853, 1069, 1361, 1709, 2137, 2677, 3347, 4201, 5261, 6577, 8231, 10289,
                12889, 16127, 20161, 25219, 31531, 39419, 49277, 61603, 77017, 96281,
                120371, 150473, 188107, 235159, 293957, 367453, 459317, 574157, 717697,
```

**Comment by widlarizer** [comment]
@povik the prime list change adds runtime improvements up to 3% on top of this with a 0.5% regression in one case

File: backends/cxxrtl/cxxrtl_backend.cc (hunk 8 above)
Comment (widlarizer) [reply backends/cxxrtl/cxxrtl_backend.cc:867]: Yep

File: frontends/verific/verific.cc [hunk 10]
```diff
@@ -238,19 +238,7 @@ RTLIL::IdString VerificImporter::new_verific_id(Verific::DesignObj *obj)
 
 RTLIL::Const mkconst_str(const std::string &str)
```
Comment (povik) [inline frontends/verific/verific.cc:239]: We should remove this helper altogether

File: kernel/rtlil.cc (hunk 6 above)
Comment (povik) [reply kernel/rtlil.cc]: Still applies

File: kernel/rtlil.cc [hunk 11]
```diff
@@ -203,93 +203,205 @@ const pool<IdString> &RTLIL::builtin_ff_cell_types() {
...
-	if (bits.size() != other.bits.size())
-		return bits.size() < other.bits.size();
-	for (size_t i = 0; i < bits.size(); i++)
-		if (bits[i] != other.bits[i])
-			return bits[i] < other.bits[i];
+	if (is_str() && other.is_str())
+		return str_ < other.str_;
+
+	// If either operand isn't a string, we have to construct a copy
+	// to bitvectorize without unpacking either operand
```
Comment (povik) [inline kernel/rtlil.cc:308]: This would be solved if we had a constant bit iterator which didn't unpack

File: kernel/rtlil.h [hunk 12]
```diff
@@ -22,6 +22,8 @@
 
 #include "kernel/yosys_common.h"
 #include "kernel/yosys.h"
+#include <variant>
```
Comment (povik) [inline kernel/rtlil.h:25]: Include no longer needed

File: kernel/rtlil.h [hunk 13]
```diff
@@ -657,38 +667,111 @@ namespace RTLIL
...
+	// Do not access the union or tag even in Const methods unless necessary
+	mutable backing_tag tag;
+	union {
+		mutable bitvectype bits_;
+		mutable std::string str_;
+	};
+
+	// Use these utilities instead
+	bool is_bits() const { return tag == backing_tag::bits; }
+	bool is_str() const { return tag == backing_tag::string; }
```
Comment (povik) [inline kernel/rtlil.h:683]: Mark those private, possibly rename those, otherwise someone at one point will take those to be semantically significant (e.g. mistake them for a `flags & CONST_FLAG_STRING != 0` check)

I suggest `bits_backed()` and `str_backed()`

File: kernel/rtlil.h [hunk 14]
```diff
@@ -657,38 +667,111 @@ namespace RTLIL
...
 
-	Const() : flags(RTLIL::CONST_FLAG_NONE) {}
+	bitvectype* get_if_bits() const { return is_bits() ? &bits_ : NULL; }
+	std::string* get_if_str() const { return is_str() ? &str_ : NULL; }
+
+	void check(bool condition) const {
+		// TODO debug assert or throw
+		// log_assert(condition && "malformed Const union");
+	}
```
Comment (povik) [inline kernel/rtlil.h:691]: I am not convinced we want this helper. This is worse than `log_assert` at the call site unless we start inspecting backtrace here

File: kernel/rtlil.h [hunk 15]
```diff
@@ -657,38 +667,111 @@ namespace RTLIL
...
 	bool operator <(const RTLIL::Const &other) const;
 	bool operator ==(const RTLIL::Const &other) const;
 	bool operator !=(const RTLIL::Const &other) const;
 
+	const std::vector<RTLIL::State>& bits() const;
+	std::vector<RTLIL::State>& bits();
 	bool as_bool() const;
 	int as_int(bool is_signed = false) const;
-	std::string as_string() const;
+	std::string as_string(std::string any = "-") const;
```
Comment (povik) [inline kernel/rtlil.h:715]: Better served with a `const char *` argument I think

File: passes/cmds/setundef.cc [hunk 16]
```diff
@@ -243,6 +243,8 @@ struct SetundefPass : public Pass {
 			{
 				for (auto *cell : module->selected_cells()) {
 					for (auto &parameter : cell->parameters) {
+						if (parameter.second.flags & RTLIL::CONST_FLAG_STRING)
+							continue;
```
Comment (povik) [inline passes/cmds/setundef.cc:247]: I know `parameter.second.flags & RTLIL::CONST_FLAG_STRING` implies there shouldn't be any undef bits, but this makes the code harder to read for little gain

File: kernel/rtlil.h [hunk 17]
```diff
@@ -657,38 +667,111 @@ namespace RTLIL
...
+	short flags;
+	private:
+	typedef std::vector<RTLIL::State> bitvectype;
+	enum class backing_tag: bool { bits, string };
+	// Do not access the union or tag even in Const methods unless necessary
+	mutable backing_tag tag;
+	union {
+		mutable bitvectype bits_;
+		mutable std::string str_;
+	};
```
Comment (povik) [inline kernel/rtlil.h:679]: What was the issue with `std::variant` again?

File: kernel/rtlil.cc [hunk 18]
```diff
@@ -365,6 +365,20 @@ std::vector<RTLIL::State> RTLIL::Const::to_bits() const
 	return b;
 }
 
+std::string RTLIL::Const::pretty_fmt() const {
+	if (get_if_str())
```
Comment (povik) [inline kernel/rtlil.cc:369]: This leaks the backing

File: kernel/rtlil.cc [hunk 19]
```diff
@@ -365,6 +365,20 @@ std::vector<RTLIL::State> RTLIL::Const::to_bits() const
...
+		return decode_string();
+	else
+		return std::to_string(as_int());
+}
+
+std::string RTLIL::Const::pretty_fmt_undef() const {
+	if (get_if_str())
```
Comment (povik) [inline kernel/rtlil.cc:376]: Here too

File: passes/opt/opt_merge.cc [hunk 20]
```diff
@@ -140,8 +140,13 @@ struct OptMergeWorker
 			hash_conn_strings.push_back(s + "\n");
 		}
 
-		for (auto &it : cell->parameters)
-			hash_conn_strings.push_back("P " + it.first.str() + "=" + it.second.as_string() + "\n");
```
Comment (povik) [inline passes/opt/opt_merge.cc:144]: I don't understand why we want this change. The original is more readable

File: kernel/rtlil.h [hunk 21]
```diff
@@ -715,6 +715,9 @@ struct RTLIL::Const
 	std::string as_string(std::string any = "-") const;
 	static Const from_string(const std::string &str);
 	std::vector<RTLIL::State> to_bits() const;
+	std::string pretty_fmt() const;
+	std::string pretty_fmt_undef() const;
```
Comment (povik) [inline kernel/rtlil.h:719]: These helpers both have a single call site, it's not obvious what they do, I suggest we remove them

File: kernel/rtlil.cc (hunk 6 above)
Comment (povik) [reply kernel/rtlil.cc]: Overridden in a subsequent commit

File: kernel/rtlil.h (hunk 13 above)
Comment (widlarizer) [reply kernel/rtlil.h:683]: They are private

Comment (povik) [reply kernel/rtlil.h:683]: Ah the indentation threw me off, we should fix it

File: kernel/rtlil.h (hunk 14 above)
Comment (widlarizer) [reply kernel/rtlil.h:691]: How about this hack:

```diff
+#define check(condition) log_assert(condition && "malformed Const union")
 struct RTLIL::Const
 {
        short flags;
@@ -685,11 +686,6 @@ struct RTLIL::Const
        bitvectype* get_if_bits() const { return is_bits() ? &bits_ : NULL; }
        std::string* get_if_str() const { return is_str() ? &str_ : NULL; }

-       void check(bool condition) const {
-               // TODO debug assert or throw
-               // log_assert(condition && "malformed Const union");
-       }
-
        bitvectype& get_bits() const { check(is_bits()); return *get_if_bits(); }
        std::string& get_str() const { check(is_str()); return *get_if_str(); }
        public:
@@ -816,6 +812,7 @@ struct RTLIL::Const
                return h;
        }
 };
+#undef check
```

Thoughts? I previously verified with benchmarks that keeping the assert is free. Those asserts will make everything so verbose

Comment (povik) [reply kernel/rtlil.h:691]: I'm fine with it

File: passes/cmds/setundef.cc (hunk 16 above)
Comment (widlarizer) [reply passes/cmds/setundef.cc:247]: If this isn't done, then bits() is iterated over. In my current implementation, that means unpacking. I don't think this is a readability problem, maybe I'm blind to that with things I wrote myself, but "skip these because they're not interesting anyway" is a common pattern that's easy to recognize for me

Comment (povik) [reply passes/cmds/setundef.cc:247]: Yes, but we don't care about unpacking for parameters

File: passes/opt/opt_merge.cc (hunk 20 above)
Comment (widlarizer) [reply passes/opt/opt_merge.cc:144]: This is a functional change that hashes fewer characters. However, since we're not concerned with string parameters, only string attributes, and this only works on parameters, the change wouldn't be used, so I guess I'll put it back as you suggest

File: kernel/rtlil.cc (hunk 18 above)
Comment (widlarizer) [reply kernel/rtlil.cc:369]: Yes, it's used for pretty printing, which isn't supposed to be some kind of correctness-dependent process, but as a compact string representation readable by a human

File: kernel/rtlil.cc (hunk 11 above)
Comment (widlarizer) [reply kernel/rtlil.cc:308]: Do you expect a runtime or readability improvement from having two iterators? Or do you think it's okay to never unpack?

File: kernel/rtlil.cc (hunk 18 above)
Comment (povik) [reply kernel/rtlil.cc:369]: There's no guarantee someone will not be relying on the output not changing if the underlying value hasn't changed. Case in point: you have used it this way for hashing in opt_merge.cc

Comment (widlarizer) [reply kernel/rtlil.cc:369]: Good point, I didn't consider that the same string packed and unpacked then has to be equal in hashing input as well

File: kernel/rtlil.cc (hunk 11 above)
Comment (povik) [reply kernel/rtlil.cc:308]: I think it's okay ot never unpack in the comparison operator, but I feel better about having a non-unpacking constant iterator, and using it to implement all operations where we care about not inadvertently unpacking, than about constructing temporary copies which we let unpack. There may be a minimal performance benefit too.

The constant iterator would also solve the semantical question, if comparisons were implemented on top of the same iterator no matter the backing.

Comment (widlarizer) [reply kernel/rtlil.cc:308]: as discussed elsewhere, I'm adding a const iterator that doesn't unpack

File: kernel/rtlil.cc [hunk 22]
```diff
@@ -203,93 +203,198 @@ const pool<IdString> &RTLIL::builtin_ff_cell_types() {
...
+
+bool RTLIL::Const::operator<(const RTLIL::Const &other) const
 {
-	if (bits.size() != other.bits.size())
-		return bits.size() < other.bits.size();
-	for (size_t i = 0; i < bits.size(); i++)
-		if (bits[i] != other.bits[i])
-			return bits[i] < other.bits[i];
+	if (is_str() && other.is_str())
+		return str_ < other.str_;
```
Comment (povik) [inline kernel/rtlil.cc:305]: The comparison still leaks backing

File: kernel/rtlil.h [hunk 23]
```diff
@@ -657,38 +665,205 @@ namespace RTLIL
...
 	bool operator ==(const RTLIL::Const &other) const;
 	bool operator !=(const RTLIL::Const &other) const;
 
+	std::vector<RTLIL::State>& bits();
 	bool as_bool() const;
 	int as_int(bool is_signed = false) const;
-	std::string as_string() const;
+	std::string as_string(const char* any = "-") const;
 	static Const from_string(const std::string &str);
+	std::vector<RTLIL::State> to_bits() const;
```
Comment (povik) [inline kernel/rtlil.h:715]: We have `to_bits()` and `bits().to_vec()` which AFAICT do the same thing

File: kernel/rtlil.cc [hunk 24]
```diff
@@ -203,93 +203,198 @@ const pool<IdString> &RTLIL::builtin_ff_cell_types() {
...
+
+std::vector<RTLIL::State> RTLIL::Const::to_bits() const
+{
+	if (auto bv = get_if_bits()) {
+		return *bv;
+	}
+	auto& str = get_str();
+	bitvectype b;
+	b.reserve(str.size() * 8);
+	for (int i = str.size()-1; i >= 0; i--) {
```
Comment (povik) [inline kernel/rtlil.cc:351]: This could be implemented on top of the constant iterator

File: kernel/rtlil.h [hunk 25]
```diff
@@ -657,38 +665,205 @@ namespace RTLIL
...
+			using pointer = const State*;
+			using reference = State;
+
+			const_iterator(const Const& c, size_t i) : parent(c), idx(i) {}
+
+			State operator*() const {
+				if (auto bv = parent.get_if_bits())
+					return (*bv)[idx];
+
+				bool bit = (parent.get_str()[idx / 8] & (1 << (7 - (idx % 8))));
```
Comment (povik) [inline kernel/rtlil.h:798]: The indexing here is reversed both for the character selection and the bit in a character

File: kernel/rtlil.h [hunk 26]
```diff
@@ -657,38 +665,205 @@ namespace RTLIL
...
+
+			const_iterator operator+(int add) {
+				return const_iterator(parent, idx + add);
+			}
+			int operator-(const const_iterator& other) {
+				return idx - other.idx;
+			}
+
+			bool operator==(const const_iterator& other) const {
+				return idx == other.idx;
```
Comment (povik) [inline kernel/rtlil.h:830]: `&& &parent == &other.parent`, or are comparisons on iterators from different parents undefined? Not sure here what is mandated

File: kernel/rtlil.h [hunk 27]
```diff
@@ -657,35 +665,205 @@ namespace RTLIL
...
+		}
+		std::vector<State> to_vec() const {
+			std::vector<State> v;
+			for (auto bit : *this)
+				v.push_back(bit);
+			return v;
+		}
+	};
+
+	const const_bit_view bits() const {
```
Comment (povik) [inline kernel/rtlil.h:864]: I'm a bit worried naming the methods the same will lead to confusion. What about not having the `bits() const` method at all, and instead retain `begin()` and `end()` only? This will mean there's only one way to iterate over a Const object, which is `for (State bit : constval)`

File: kernel/rtlil.h (hunk 25 above)
Comment (widlarizer) [reply kernel/rtlil.h:798]: Strange, I actually tested it before, but very manually

File: kernel/rtlil.h (hunk 27 above)
Comment (widlarizer) [reply kernel/rtlil.h:864]: 1. What do you mean by retain? Const doesn't have a `begin()` method on main
2. You mean we should remove the iterator that can mutate Const? I'm pretty sure there's code that relies on that

Comment (povik) [reply kernel/rtlil.h:864]: > 1. What do you mean by retain? Const doesn't have a `begin()` method on main

Not on main but in this PR it does.

> 2\. You mean we should remove the iterator that can mutate Const? I'm pretty sure there's code that relies on that

No, the other one, basically get rid of `const_bit_view` but keep `begin()` and `end()` on the Const itself as the way of accessing what is now `const_bit_view::const_iterator`

Comment (widlarizer) [reply kernel/rtlil.h:864]: So there will be two ways of iterating Const. One is `(auto a : c)` and the other is `(auto a : c.bits())`. Both give you the same values, but the second one is mutable and makes performance worse? It's doable, I suppose

Comment (povik) [reply kernel/rtlil.h:864]: Yes, that's what I am proposing

File: kernel/rtlil.cc [hunk 28]
```diff
@@ -203,93 +203,157 @@ const pool<IdString> &RTLIL::builtin_ff_cell_types() {
...
 bool RTLIL::Const::operator <(const RTLIL::Const &other) const
 {
-	if (bits.size() != other.bits.size())
-		return bits.size() < other.bits.size();
-	for (size_t i = 0; i < bits.size(); i++)
-		if (bits[i] != other.bits[i])
-			return bits[i] < other.bits[i];
+	const char* ctx = "operator<";
+	if (std::get_if<std::string>(&backing) != std::get_if<std::string>(&other.backing))
+		return decode_string() < other.decode_string();
```
[the remark this answers is not in the document]
Comment (widlarizer) [reply kernel/rtlil.cc]: Should be fixed now

File: frontends/ast/ast.cc [hunk 29]
```diff
@@ -1749,16 +1741,7 @@ static std::string serialize_param_value(const RTLIL::Const &val) {
 		res.push_back('r');
 	res += stringf("%d", GetSize(val));
 	res.push_back('\'');
-	for (int i = GetSize(val) - 1; i >= 0; i--) {
```
[the remark this answers is not in the document]
Comment (widlarizer) [reply frontends/ast/ast.cc:1752]: I still think refactoring into as_string is the correct thing to do when we have the opportunity

**Review by povik — COMMENTED** [review]
Looks good apart from the extra cxxrtl change and the extra assert

File: kernel/rtlil.cc [hunk 30]
```diff
@@ -200,96 +200,196 @@ const pool<IdString> &RTLIL::builtin_ff_cell_types() {
...
 
 bool RTLIL::Const::operator !=(const RTLIL::Const &other) const
 {
-	return bits != other.bits;
+	return !(*this == other);
+}
+
+std::vector<RTLIL::State>& RTLIL::Const::bits()
+{
+	log_assert(is_bits());
```
[the remark this answers is not in the document]
Comment (widlarizer) [reply kernel/rtlil.cc:343]: It's redundant now, but it wasn't before

File: kernel/rtlil.cc [hunk 31]
```diff
@@ -3730,7 +3913,7 @@ RTLIL::SigChunk::SigChunk(const RTLIL::SigBit &bit)
 	wire = bit.wire;
 	offset = 0;
 	if (wire == NULL)
-		data = RTLIL::Const(bit.data).bits;
+		data = RTLIL::Const(bit.data).to_bits();
```
[the remark this answers is not in the document]
Comment (widlarizer) [reply kernel/rtlil.cc:3916]: nice catch

File: kernel/rtlil.cc (hunk 30 above)
[the remark this answers is not in the document]
Comment (povik) [reply kernel/rtlil.cc:343]: As I understand it would crash on the following
```
Const c("blah"s);
c.bits();
```

[the remark this answers is not in the document]
Comment (widlarizer) [reply kernel/rtlil.cc:343]: true, I misread the called code

**Review by povik — APPROVED** [review]
I haven't looked at the new unit testing but the Const change itself looks fine with the points from the last review resolved

**Review by whitequark — COMMENTED** [review]
Great work!

**Comment by widlarizer** [comment]
just some fixes from the functional merges

**Comment by widlarizer** [comment]
I validated the memory savings to be 0-30%.

ORFS peak:

| design | 5c9b2df68 main [MB] | 2dd0ae31f PR [MB] |
| --- | --- | --- |
|asap7_ibex|224.52|194.96|
|sky130hd_gcd|57.75|57.75|
|sky130hd_ibex|223.20|159.32|
|sky130hd_jpeg|726.52|613.00|

In the flow I dumped out the RTLIL before synthesis and outside of ORFS loaded the RTLIL and ran just the basic `synth` command:

| design | 5c9b2df68 main [MB] | 2dd0ae31f PR [MB] |
| --- | --- | --- |
|asap7_ibex|175.88|122.42|
|sky130hd_gcd|45.56|45.56|
|sky130hd_ibex|175.76|123.02|
|sky130hd_jpeg|85.14|75.63|

As for runtime, there was negligible improvement (0-1%)

**Comment by KrystalDelusion** [comment]
This PR introduced a lot of warnings along the lines of
`comparison of integers of different signs: 'size_t' (aka 'unsigned long') and 'int' [-Wsign-compare]`, I'm guessing because the previous method of checking string length used the `.size()` method of a `std::vector` which returns a `size_t`, but the `RTLIL::const` `.size()` method returns it as an `int` (and seems to implicitly convert the underlying `.size()` `size_t` value)

**Comment by widlarizer** [comment]
True. I brought up size types on a dev jf and we decided to stick with int for Yosys data structures so I guess I'll make `Const::size()` consistent with that then
